internal/contributebot/ghauth.go

```diff
@@ -0,0 +1,308 @@
+// Copyright 2018 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"crypto/rsa"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/google/go-cloud/runtimevar"
+	"github.com/google/go-github/github"
+)
+
+// gitHubAppAuth makes HTTP requests with GitHub application credentials.
+type gitHubAppAuth struct {
+	id   int64
+	base http.RoundTripper
+
+	available   <-chan struct{}
+	cancelWatch context.CancelFunc
+	watchDone   <-chan struct{}
+
+	mu         sync.RWMutex
+	privateKey *rsa.PrivateKey
+}
+
+func newGitHubAppAuth(id int64, privateKey *runtimevar.Variable, rt http.RoundTripper) *gitHubAppAuth {
+	done := make(chan struct{})
+	available := make(chan struct{})
+	ctx, cancel := context.WithCancel(context.Background())
+	auth := &gitHubAppAuth{
+		id:          id,
+		base:        rt,
+		available:   available,
+		cancelWatch: cancel,
+		watchDone:   done,
+	}
+	go func() {
+		defer close(done)
```

Comment:
Maybe a comment like:

// Watches privateKey until Stop is called to cancel ctx.